HW2/P5/physics.pyx

```diff
@@ -69,31 +83,53 @@ cdef void sub_update(FLOAT[:, ::1] XY,
     ############################################################
     # SUBPROBLEM 2: use the grid values to reduce the number of other
     # objects to check for collisions.
-    for j in range(i + 1, count):
-        XY2 = &(XY[j, 0])
-        V2 = &(V[j, 0])
-        if overlapping(XY1, XY2, R):
-            # SUBPROBLEM 4: Add locking
-            if not moving_apart(XY1, V1, XY2, V2):
-                collide(XY1, V1, XY2, V2)
-
-            # give a slight impulse to help separate them
-            for dim in range(2):
-                V2[dim] += eps * (XY2[dim] - XY1[dim])
+    grid_x = <unsigned int>(XY[i,0] / grid_spacing)
+    grid_y = <unsigned int>(XY[i,1] / grid_spacing)
+    grid_size = <unsigned int>((1.0 / grid_spacing) + 1)
+
+    #We check all the points in the grid that is 2 spaces away from the ball
+    for x in range(grid_x + 1, grid_x + 3):
+        for y in range(grid_y + 1, grid_y + 4 + grid_x - x):
+            #check if x and y are bound within the [0,1]^2 space, and check if the grid value is smaller than count
+            if x < grid_size and y < grid_size and Grid[x,y] < count:
+                XY2 = &(XY[Grid[x,y], 0])
+                V2 = &(V[Grid[x,y], 0])
+                if overlapping(XY1, XY2, R):
+                    # SUBPROBLEM 4: Add locking
+
+                    #Acquire lock for XY2
+                    acquire(&(locks[Grid[x,y]]))
+
+                    if not moving_apart(XY1, V1, XY2, V2):
+                        collide(XY1, V1, XY2, V2)
+
+                    # give a slight impulse to help separate them
+                    for dim in range(2):
+                        V2[dim] += eps * (XY2[dim] - XY1[dim])
+
+                    #Release lock for XY2
+                    release(&(locks[Grid[x,y]]))
+    release(&(locks[i]))
```

Comment:
I think you should compare the index too. If not, balls will collide twice and there will be deadlock.